core/templates/services/classifier-data-backend-api.service.ts

```diff
@@ -0,0 +1,140 @@
+  private readonly classifierDataDownloadUrlTemplate: string;
+
+  constructor(
+      private http: HttpClient,
+      private urlInterpolationService: UrlInterpolationService) {
+    if (
+      !ClassifierDataBackendApiService.DEV_MODE &&
+      !ClassifierDataBackendApiService.GCS_RESOURCE_BUCKET_NAME) {
+      throw new Error('GCS_RESOURCE_BUCKET_NAME is not set in prod.');
+    }
+
+    const urlPrefix = ClassifierDataBackendApiService.DEV_MODE ?
+      '/_ah/gcs/' : 'https://storage.googleapis.com/';
+    this.classifierDataDownloadUrlTemplate = (
+      urlPrefix + ClassifierDataBackendApiService.GCS_RESOURCE_BUCKET_NAME +
+      '/<entity_type>/<entity_id>/assets/<filename>');
+  }
+
+  static get DEV_MODE(): boolean {
+    return AppConstants.DEV_MODE;
+  }
+
+  static get GCS_RESOURCE_BUCKET_NAME(): string {
+    return ClassifierDataBackendApiService.DEV_MODE ?
+      AppConstants.DEFAULT_GCS_RESOURCE_BUCKET_NAME :
+      AppConstants.GCS_RESOURCE_BUCKET_NAME;
+  }
+
+  private getDownloadUrl(
+      entityType: string, entityId: string, filename: string): string {
+    return this.urlInterpolationService.interpolateUrl(
+      this.classifierDataDownloadUrlTemplate, {
+        entity_type: entityType,
+        entity_id: entityId,
+        filename: filename,
+      });
+  }
+
+  private async getClassifierMetaData(
+      explorationId: string, explorationVersion: number,
+      stateName: string): Promise<ClassifierMetaData> {
+    return new Promise((resolve, reject) => {
+      this.http.get<ClassifierMetaDataBackendDict>(
+        '/ml/trainedclassifierhandler', {
+          params: {
+            exploration_id: explorationId,
+            exploration_version: explorationVersion.toString(),
+            state_name: stateName
+          },
+          responseType: 'json'
+        }).toPromise().then(response => {
+        resolve({
+          algorithmId: response.algorithm_id,
+          algorithmVersion: response.algorithm_version,
+          filename: response.gcs_filename
+        });
+      }, errorResponse => {
+        reject(errorResponse);
+      });
+    });
+  }
+
+  async getClassifierData(
+      explorationId: string, explorationVersion: number,
+      stateName: string): Promise<Classifier> {
+    return new Promise((resolve, reject) => {
+      this.getClassifierMetaData(
+        explorationId, explorationVersion, stateName).then(
+        response => {
+          let classifierMetaData = response;
+          this.http.get(
+            this.getDownloadUrl(
+              AppConstants.ENTITY_TYPE.EXPLORATION, explorationId,
+              response.filename), {
+              responseType: 'arraybuffer'
+            }).toPromise().then(response => {
+            resolve(new Classifier(
+              classifierMetaData.algorithmId,
+              unzipSync(Buffer.from(response)),
+              classifierMetaData.algorithmVersion
+            ));
+          }, errorResponse => {
```

Comment:
There are two different errorResponses being dealt with here, perhaps name them differently to disambiguate.

Reply:
Done